Userland/Libraries/LibWeb/FileAPI/Blob.cpp

```diff
@@ -0,0 +1,163 @@
+#include <LibWeb/FileAPI/Blob.h>
+
+namespace Web::FileAPI {
+
+// https://w3c.github.io/FileAPI/#ref-for-dom-blob-blob
+Blob::Blob(Optional<Vector<BlobPart>> const& blob_parts, Optional<BlobPropertyBag> const& options)
+{
+    // 1. If invoked with zero parameters, return a new Blob object consisting of 0 bytes, with size set to 0, and with type set to the empty string.
+    if (!blob_parts.has_value() && !options.has_value())
+        return;
+
+    // 2. Let bytes be the result of processing blob parts given blobParts and options.
+    if (blob_parts.has_value())
+        m_byte_buffer = process_blob_parts(blob_parts.value());
+
+    if (options.has_value())
+        m_type = options->type;
+}
+
+Blob::Blob(ByteBuffer const& byte_buffer, String const& type)
+    : m_byte_buffer(move(byte_buffer))
+    , m_type(move(type))
+{
+}
+
+// https://w3c.github.io/FileAPI/#process-blob-parts
+ByteBuffer Blob::process_blob_parts(Vector<BlobPart> const& blob_parts)
+{
+    // 1. Let bytes be an empty sequence of bytes.
+    ByteBuffer bytes {};
+
+    // 2. For each element in parts:
+    for (auto const& blob_part : blob_parts) {
+        // 1. If element is a USVString, run the following substeps:
+        if (blob_part.has<String>()) {
+            // 1. Let s be element.
+            auto s = blob_part.get<String>();
+
+            // FIXME: 2. If the endings member of options is "native", set s to the result of converting line endings to native of element.
+
+            // 3. Append the result of UTF-8 encoding s to bytes.
+            bytes.append(s.to_byte_buffer());
+        }
+
+        // FIXME: 2. If element is a BufferSource, get a copy of the bytes held by the buffer source, and append those bytes to bytes.
+
+        // 3. If element is a Blob, append the bytes it represents to bytes.
+        if (blob_part.has<NonnullRefPtr<Blob>>()) {
+            auto blob = blob_part.get<NonnullRefPtr<Blob>>();
+            bytes.append(blob->m_byte_buffer.bytes());
+        }
+    }
+    return bytes;
+}
+
+// https://w3c.github.io/FileAPI/#dfn-slice
+NonnullRefPtr<Blob> Blob::slice(Optional<i64> start, Optional<i64> end, Optional<String> const& content_type)
+{
+    // 1. The optional start parameter is a value for the start point of a slice() call, and must
+    // be treated as a byte-order position, with the zeroth position representing the first byte.
+    // User agents must process slice() with start normalized according to the following:
+    i64 relative_start;
+    if (!start.has_value()) {
+        // a. If the optional start parameter is not used as a parameter when making this call, let relativeStart be 0.
+        relative_start = 0;
+    } else {
+        auto start_value = start.value();
+        // b. If start is negative, let relativeStart be max((size + start), 0).
+        if (start_value < 0)
+            relative_start = max((size() + start_value), 0);
+        // c. Else, let relativeStart be min(start, size).
```

Comment:
This looks really weird, maybe add braces?

Reply:
I don't think adding braces adds any value to the weirdness. It's the same pattern as used here: https://github.com/SerenityOS/serenity/blob/3f93aec7207b646cf7b125a86451a028fc8e766a/Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp#L2037

Reply:
I'm ok with this style, but if it makes CxByte happy I don't mind the curly braces either. (In general I'm much more happy with curlies around single-line blocks now, given that they will be mandatory in Jakt code)

Reply:
Then we need to put the comment inside the `else`, and not above, otherwise formatting will indent it one step

Reply:
Added braces

Reply:
> Then we need to put the comment inside the else, not above, otherwise formatting will indent it one step

Hm? The new style is backwards as it has code before instructions, what's wrong with this?

https://github.com/SerenityOS/serenity/blob/fec5d8d5310000fc95cfffa7d1ef20ecd0d29c01/Userland/Libraries/LibJS/Runtime/Temporal/PlainDate.cpp#L246-L261

Reply:
Just me brain not worky. I'll fix it up according to that on my next push